src/Bicep.Core/Intermediate/Expression.cs

```diff
@@ -492,13 +492,13 @@ public record DeclaredResourceExpression(
     SyntaxBase BodySyntax,
     Expression Body,
     ImmutableArray<ResourceDependencyExpression> DependsOn,
-    Expression? Description = null,
-    ArrayExpression? RetryOn = null,
-    ArrayExpression? WaitUntil = null
+    Expression? Description = null
 ) : DescribableExpression(SourceSyntax, Description)
 {
     public override void Accept(IExpressionVisitor visitor)
         => visitor.VisitDeclaredResourceExpression(this);
+
+    public Dictionary<string, ArrayExpression> DecoratorConfig { get; } = new Dictionary<string, ArrayExpression>();
```

Comment:
Can this be moved into the record constructor, and made immutable?

```suggestion
    Expression? Description = null,
    ImmutableDictionary<string, ArrayExpression>? DecoratorConfig = null
) : DescribableExpression(SourceSyntax, Description)
{
    public override void Accept(IExpressionVisitor visitor)
        => visitor.VisitDeclaredResourceExpression(this);
```

Reply:
Working with records & immutable structures takes a bit of getting used to. To give an example, your code in SystemNamespaceType is currently:

```csharp
if (decorated is DeclaredResourceExpression declaredResourceExpression)
{
    declaredResourceExpression.DecoratorConfig.Add("OnlyIfNotExists", new ArrayExpression(null, []));
    return declaredResourceExpression;
}
```

If using my above suggestion, this could be rewritten as:
```csharp
if (decorated is DeclaredResourceExpression resource)
{
    return resource with {
        DecoratorConfig = resource.DecoratorConfig.Add("OnlyIfNotExists", new ArrayExpression(null, []))
    };
}
```

Reply:
Yes, will update. I had moved the Dictionary out of the record constructor to avoid having to do null checks when adding the decorator options. 

```C#
if (decorated is DeclaredResourceExpression declaredResourceExpression)
{
    return declaredResourceExpression with
    {
        DecoratorConfig = declaredResourceExpression.DecoratorConfig?.Add("OnlyIfNotExists", new ArrayExpression(null, []))
            ?? ImmutableDictionary<string, ArrayExpression>.Empty.Add("OnlyIfNotExists", new ArrayExpression(null, []))
    };
}
```